rust/datafusion/src/optimizer/type_coercion.rs

```diff
@@ -43,138 +45,77 @@ impl<'a> TypeCoercionRule<'a> {
-            Expr::Cast { .. } => Ok(expr.clone()),
-            Expr::Column(_) => Ok(expr.clone()),
-            Expr::Alias(expr, alias) => Ok(Expr::Alias(
-                Box::new(self.rewrite_expr(expr, schema)?),
-                alias.to_owned(),
-            )),
-            Expr::Literal(_) => Ok(expr.clone()),
-            Expr::Not(_) => Ok(expr.clone()),
-            Expr::Sort { .. } => Ok(expr.clone()),
-            Expr::Wildcard { .. } => Err(ExecutionError::General(
-                "Wildcard expressions are not valid in a logical query plan".to_owned(),
-            )),
-            Expr::Nested(e) => self.rewrite_expr(e, schema),
-        }
+            _ => {}
+        };
+        utils::from_expression(expr, &expressions)
     }
 }
 
 impl<'a> OptimizerRule for TypeCoercionRule<'a> {
     fn optimize(&mut self, plan: &LogicalPlan) -> Result<LogicalPlan> {
-        match plan {
-            LogicalPlan::Projection { expr, input, .. } => {
-                LogicalPlanBuilder::from(&self.optimize(input)?)
-                    .project(self.rewrite_expr_list(expr, input.schema())?)?
-                    .build()
-            }
-            LogicalPlan::Selection { expr, input, .. } => {
-                LogicalPlanBuilder::from(&self.optimize(input)?)
-                    .filter(self.rewrite_expr(expr, input.schema())?)?
-                    .build()
-            }
-            LogicalPlan::Aggregate {
-                input,
-                group_expr,
-                aggr_expr,
-                ..
-            } => LogicalPlanBuilder::from(&self.optimize(input)?)
-                .aggregate(
-                    self.rewrite_expr_list(group_expr, input.schema())?,
-                    self.rewrite_expr_list(aggr_expr, input.schema())?,
-                )?
-                .build(),
-            LogicalPlan::TableScan { .. } => Ok(plan.clone()),
-            LogicalPlan::InMemoryScan { .. } => Ok(plan.clone()),
-            LogicalPlan::ParquetScan { .. } => Ok(plan.clone()),
-            LogicalPlan::CsvScan { .. } => Ok(plan.clone()),
-            LogicalPlan::EmptyRelation { .. } => Ok(plan.clone()),
-            LogicalPlan::Limit { .. } => Ok(plan.clone()),
-            LogicalPlan::Sort { .. } => Ok(plan.clone()),
-            LogicalPlan::CreateExternalTable { .. } => Ok(plan.clone()),
-        }
+        let inputs = utils::inputs(plan);
+        let expressions = utils::expressions(plan);
+
+        // apply the optimization to all inputs of the plan
+        let new_inputs = inputs
+            .iter()
+            .map(|plan| self.optimize(*plan))
+            .collect::<Result<Vec<_>>>()?;
+        // re-write all expressions on this plan.
+        // This assumes a single input, [0]. It wont work for join, subqueries and union operations with more than one input.
+        // It is currently not an issue as we do not have any plan with more than one input.
+        let new_expressions = expressions
```

Comment:
Good catch. No because I am unsure how that could be possible: if we have expressions on a plan, we need an `input` to convert them to physical expressions and evaluate them against. AFAIK an expression always requires an input to be evaluated against. 

Do you have an example in mind?

AFAIK even a literal expression requires a schema to pass to `Expr::get_type` and `Expr::name`.

Reply:
```suggestion
        assert!(expressions.len() == 0 || inputs.len() > 0, "Assume that all plan nodes with expressions had inputs");
        let new_expressions = expressions
```

I think the `EmptyRelation`,  https://github.com/apache/arrow/blob/master/rust/datafusion/src/logicalplan.rs#L761-L764, for example has no input LogicalPlan, but perhaps you are saying "even though `EmptyRelation` has no inputs (and thus could cause `inputs[0].schema()` to panic) it also has no Expressions then the potential panic'ing code won't be run. 

I guess I was thinking to the  future where we add expressions to root nodes (e.g. perhaps filtering *during* a table scan or something) which would then have expressions but no input.

 I think this code is fine as is. Perhaps we could make the code slightly easier to work with in the future if we did something like the assert suggestion here that there were no inputs if there were expressions rather than panic.